From 9c346b1bf29b8cd3b3db5c37e769020d7d6fcc3f Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:18 -0400
Subject: [virt] kvm/nvmx: Allow nested guests to run with dirty debug registers

Message-id: <1406040016-3289-164-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86242
O-Subject: [RHEL7 PATCH v2 163/221] KVM: nVMX: Allow nested guests to run with dirty debug registers
Bugzilla: 1068627 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

When preparing the VMCS02, the CPU-based execution controls is computed
by vmx_exec_control.  Turn off DR access exits there, too, if the
KVM_DEBUGREG_WONT_EXIT bit is set in switch_db_regs.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit d16c293e4ecbddedfc1d64095ce56f0569adc12b)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index a9940ec..f4e5aed 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -4242,6 +4242,10 @@ static u32 vmx_pin_based_exec_ctrl(struct vcpu_vmx *vmx)
 static u32 vmx_exec_control(struct vcpu_vmx *vmx)
 {
  u32 exec_control = vmcs_config.cpu_based_exec_ctrl;
+
+ if (vmx->vcpu.arch.switch_db_regs & KVM_DEBUGREG_WONT_EXIT)
+  exec_control &= ~CPU_BASED_MOV_DR_EXITING;
+
  if (!vm_need_tpr_shadow(vmx->vcpu.kvm)) {
   exec_control &= ~CPU_BASED_TPR_SHADOW;
 #ifdef CONFIG_X86_64
-- 
1.7.1